src/auth/src/credentials/internal/sts_exchange.rs

```diff
@@ -0,0 +1,442 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::credentials::errors::CredentialsError;
+use base64::Engine;
+use serde::Deserialize;
+use std::collections::HashMap;
+
+#[allow(dead_code)]
+type Result<T> = std::result::Result<T, CredentialsError>;
+
+/// Token Exchange grant type for a sts exchange.
+#[allow(dead_code)]
+pub const TOKEN_EXCHANGE_GRANT_TYPE: &str = "urn:ietf:params:oauth:grant-type:token-exchange";
+/// Refresh Token Exchange grant type for a sts exchange.
+#[allow(dead_code)]
+pub const REFRESH_TOKEN_GRANT_TYPE: &str = "refresh_token";
+/// TokenType for a sts exchange.
+#[allow(dead_code)]
+pub const ACCESS_TOKEN_TYPE: &str = "urn:ietf:params:oauth:token-type:access_token";
+/// JWT TokenType for a sts exchange.
+#[allow(dead_code)]
+pub const JWT_TOKEN_TYPE: &str = "urn:ietf:params:oauth:token-type:jwt";
+
+/// Handles OAuth2 Secure Token Service (STS) exchange.
+/// Reference: https://datatracker.ietf.org/doc/html/rfc8693
+pub struct STSHandler {
+    client: reqwest::Client,
+}
+
+#[allow(dead_code)]
+impl STSHandler {
+    pub fn new() -> Self {
+        let client = reqwest::Client::new();
+        Self { client }
+    }
+
+    /// performs the token exchange using a refresh token flow with
+    /// the provided [RefreshAccessTokenRequest] information.
+    pub async fn refresh_access_token(
+        &self,
+        req: RefreshAccessTokenRequest,
+    ) -> Result<TokenResponse> {
+        let mut params: HashMap<&str, String> = HashMap::new();
+        params.insert("grant_type", REFRESH_TOKEN_GRANT_TYPE.to_string());
+        params.insert("refresh_token", req.refresh_token);
```

Comment:
Consider any of these alternatives:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=c78a8edc88c7cafa5a5ca2c5c220746d